src/node_file.cc

```diff
@@ -433,6 +435,70 @@ Local<Value> BuildStatsObject(Environment* env, const uv_stat_t* s) {
   return handle_scope.Escape(stats);
 }
 
+// Used to speed up module loading.  Returns the contents of the file as
+// a string or undefined when the file cannot be opened.  The speedup
+// comes from not creating Error objects on failure.
+static void InternalModuleReadFile(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  node::Utf8Value path(env->isolate(), args[0]);
+
+  FILE* const stream = fopen(*path, "rb");
+  if (stream == nullptr) {
+    return;
+  }
+
+  std::vector<char> chars;
+  for (;;) {
+    const size_t kBlockSize = 32 << 10;
+    const size_t start = chars.size();
+    chars.resize(start + kBlockSize);
+    const size_t numchars = fread(&chars[start], 1, kBlockSize, stream);
+    if (numchars < kBlockSize) {
+      chars.resize(start + numchars);
+    }
+    if (numchars == 0) {
+      break;
+    }
+  }
+
+  fclose(stream);
+
+  size_t start = 0;
+  if (chars.size() >= 3 && 0 == memcmp(&chars[0], "\xEF\xBB\xBF", 3)) {
+    start = 3;  // Skip UTF-8 BOM.
+  }
+
+  Local<String> chars_string =
+      String::NewFromUtf8(env->isolate(),
+                          &chars[start],
+                          String::kNormalString,
+                          chars.size() - start);
+  args.GetReturnValue().Set(chars_string);
+}
+
+// Used to speed up module loading.  Returns 1 if the path refers to
+// a file, 2 when it's a directory or < 0 on error (usually -ENOENT.)
+// The speedup comes from not creating thousands of Stat and Error objects.
+static void InternalModuleStat(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  node::Utf8Value path(env->isolate(), args[0]);
+
+  uv_fs_t req;
+  int rc = uv_fs_stat(env->event_loop(), &req, *path, nullptr);
+  if (rc == 0) {
+    const uv_stat_t* const s = static_cast<const uv_stat_t*>(req.ptr);
+    rc = 1 + !!(s->st_mode & S_IFDIR);
+  }
+  uv_fs_req_cleanup(&req);
+
+  ASSERT(rc != 0);
```

Comment:
I can change it to CHECK if you want.  It's more to signal to the reader that the return value is never zero.

Actually, I'll just go ahead and remove that.  The code in lib/module.js had the concept of a cache and I planned to use 0 as a sentinel value but I ended up removing the cache because it wasn't necessary anymore.